osutil/udev/netlink/rawsockstop.go

```diff
@@ -0,0 +1,69 @@
+package netlink
+
+import (
+	"fmt"
+	"math/bits"
+	"os"
+	"syscall"
+)
+
+// RawSockStopper returns a pair of functions to manage stopping code
+// reading from a raw socket, readableOrStop blocks until
+// fd is readable or stop was called. To work properly it sets fd
+// to non-blocking mode.
+// TODO: with go 1.11+ it should be possible to just switch to setting
+// fd to non-blocking and then wrapping the socket via os.NewFile and
+// use Close to force a read to stop.
+// c.f. https://github.com/golang/go/commit/ea5825b0b64e1a017a76eac0ad734e11ff557c8e
+func RawSockStopper(fd int) (readableOrStop func() (bool, error), stop func(), err error) {
+	if err := syscall.SetNonblock(fd, true); err != nil {
+		return nil, nil, err
+	}
+
+	stopR, stopW, err := os.Pipe()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// both stopR and stopW must be kept alive otherwise the corresponding
+	// file descriptors will get closed
+	readableOrStop = func() (bool, error) {
+		return stopperSelectReadable(fd, int(stopR.Fd()))
```

Comment:
Right, makes sense and is way simpler than I envisioned in earlier comment, thank you. I think the comment should even be expanded and made more explicit about keeping stopR within the scope of readableOrStop to prevent GC.